07_errors/jimbotech/storer_test.go

```diff
@@ -0,0 +1,116 @@
+package main
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/suite"
+)
+
+type storesSuite struct {
+	suite.Suite
+	store  Storer
+	mapper mapTest
+}
+
+const brown = "brown"
+const black = "black"
+const grey = "grey"
+
+func TestSuite(t *testing.T) {
+	suite.Run(t, &storesSuite{store: NewMapStore()})
+	suite.Run(t, &storesSuite{store: &SyncMapStore{}})
+}
+
+//SetupTest creates the correct empty map for each test
+func (s *storesSuite) SetupTest() {
+	switch s.store.(type) {
+	case MapStore:
+		s.store = NewMapStore()
+	case *SyncMapStore:
+		s.store = &SyncMapStore{}
+	default:
+		s.Fail("Unknown Storer implementation")
+	}
+	s.mapper = s.store.(mapTest)
+}
+
+func (s *storesSuite) TestReadSuccess() {
+	pup := create(s)
+	// now check by reading the value back and compare
+	pup2, err2 := s.store.ReadPuppy(pup.ID)
+	s.Require().NoError(err2)
+	s.Equal(brown, pup2.Colour)
+	// modify the retured value to make sure the
+	// value in the store does not change
+	pup2.Colour = grey
+	pup3, err2 := s.store.ReadPuppy(pup.ID)
+	s.Require().NoError(err2)
+	s.Equal(brown, pup3.Colour)
+	s.NotEqual(pup2, pup3)
+}
+
+// TestCreateSuccess add to the store and verify
+// by reading that it is in the store
+func (s *storesSuite) TestCreateSuccess() {
+	pup := create(s)
+	// Now modify the original and make sure the
+	// value in the store will not change
+	pup.Colour = black
+	// now check by reading the value back and compare
+	pup2, err2 := s.store.ReadPuppy(pup.ID)
+	s.Require().NoError(err2)
+	s.Equal("kelpie", pup2.Breed)
+	s.Equal(brown, pup2.Colour)
+	s.Equal("indispensable", pup2.Value)
+	s.True(pup2.Colour == brown)
+	s.True(pup.Colour == black)
+	s.NotEqual(pup, pup2)
+}
+
+func create(s *storesSuite) *Puppy {
+	pup := Puppy{Breed: "kelpie", Colour: brown, Value: "indispensable"}
+	id, err := s.store.CreatePuppy(&pup)
+	s.Require().NoError(err)
+	s.Require().NotEqual(pup.ID, uint32(1))
```

Comment:
this is non-obvious. Why can't `pup.ID` be `1`? I assume that this fixture function is intended to be called after a test has already created a pup? If so, at the very least this function needs a comment saying so. But I would even say this test is not needed at all - that the storer does not reuse IDs is a test on the `CreatePuppy` function and that test belongs in a test specifically for it.

[time passes as review continues]

oh wow. you generate IDs using `rand`. This makes less sense now. Why can `rand` not return `1`?

Reply:
Deleted